File: packages/singing/python/src/airi_singing_worker/dto/manifest.py
"""Manifest DTO for pipeline audit trail."""

from dataclasses import dataclass, field
from typing import Any, Dict, Optional


@dataclass
class ManifestData:
    """Data structure for the cover manifest (manifest.json)."""

    version: int = 1
    job_id: str = ""
    created_at: str = ""
    completed_at: str = ""
    separator: Dict[str, str] = field(default_factory=dict)
    pitch: Dict[str, str] = field(default_factory=dict)
    converter: Dict[str, Any] = field(default_factory=dict)
    mix: Dict[str, Any] = field(default_factory=dict)
    timing: Dict[str, float] = field(default_factory=dict)
    loudnorm: Optional[Dict[str, float]] = None
    output_sample_rate: int = 44100

File: packages/singing/src/types/manifest.ts
import type { ConverterBackendId, PitchBackendId, SeparatorBackendId } from '../constants/model-backends'
import type { PipelineStage } from '../constants/pipeline-stage'

/**
 * Full audit trail for a completed cover job.
 * Persisted as manifest.json alongside artifacts.
 */
export interface CoverManifest {
  /** Schema version for forward compatibility */
  version: 1
  jobId: string
  createdAt: string
  completedAt: string

  input: {
    originalFileName: string
    sampleRate: number
    durationSeconds: number
  }

  separator: {
    backend: SeparatorBackendId
    model: string
    leadIsolation: boolean
  }

  pitch: {
    backend: PitchBackendId
  }

  converter: {
    backend: ConverterBackendId
    model: string
    params: Record<string, unknown>
  }

  mix: {
    vocalGainDb: number
    instGainDb: number
    ducking: boolean
    targetLufs: number
    truePeakDb: number
    /** Tracks that participated in the final mix */
    tracks?: Array<{ role: string, gain: number, label: string }>
  }

  /** Per-stage timing in milliseconds */
  timing: Partial<Record<PipelineStage, number>>

  /** SHA-256 checksums for produced artifact files */
  artifactHashes?: Record<string, string>

  /** Output sample rate */
  outputSampleRate: number

  /** Post-inference evaluation results (populated when autoCalibrate is enabled) */
  evaluation?: {
    singer_similarity: number
    f0_corr: number
    source_leakage: number
    tearing_risk: number
    passed: boolean
    failed_metrics: string[]
    auto_calibrated: boolean
    params_used: {
      pitch_shift: number
      index_rate: number
      protect: number
      rms_mix_rate: number
    }
  }
}
